Docs: add docs for e2e testing (#2955)
sourcegraph/sourcegraph: doc/dev/testing.md

- To run unit tests in development (only running the tests related to uncommitted code), run `yarn test --watch`.
  - And/or use [vscode-jest](https://github.com/jest-community/vscode-jest) with `jest.autoEnable: true` (and, if you want, `jest.showCoverageOnLoad: true`)

> - To debug tests in VS Code, use [vscode-jest](https://github.com/jest-community/vscode-jest) and click the **Debug** code lens next to any `test('name ...', ...)` definition in your test file (be sure to set a breakpoint or break on uncaught exceptions by clicking in the left gutter).

Reviewer: I see the bext instructions are removed, but the e2e section below only mention the web app

Author: @felixfbecker My understanding is that bext e2e tests are pending in the PR https://github.com/sourcegraph/sourcegraph/pull/2800

@lguychard Would you recommend adding the bext to the docs here, or holding off until that PR is merged?

![image](https://user-images.githubusercontent.com/1387653/54873894-88b0c080-4d9d-11e9-9454-409ffa9864bd.png)

A video of the session is available in the **Artifacts** tab:

![image](https://user-images.githubusercontent.com/1387653/54873783-65851180-4d9b-11e9-98e5-e23b1166f2c5.png)

Here are common failure modes:

- Timed out waiting for http://localhost:7080 to be up: the `sourcegraph/server` container failed to start, so check the container logs that appear further down in the Buildkite output
- Timed out waiting for a selector to match because the CSS class in the web app changed: update the test code
- Timed out waiting for a selector to match because the page was still loading: use `waitForSelector(selector, { visible: true })`
- Page disconnected or browser session closed: another part of the test code might have called `page.close()` asynchronously, the browser crashed (check the video), or the build got canceled

> - Node was detached from the DOM: the Monaco editor changes its DOM asynchronously, so wrap interactions with it in `retry()`

Reviewer: I would say this applies to every node, not just Monaco, since React re-renders and swaps out elements async too. The solution to this should be to not reference element handles (what @beyang said).

`retry` should still be mentioned in a separate tip as the preferred way to "poll" for a condition that cannot be expressed through `waitForSelector()` (as opposed to relying on a fixed `setTimeout()`)

Drop the `-N` suffix from the name, then run:

```
gcloud container clusters get-credentials ci --zone us-central1-a --project sourcegraph-dev
kubectl port-forward -n buildkite <buildkite agent pod> 5900:5900
```

Open VNC Viewer and type in `localhost:5900`. Hit <kbd>Enter</kbd> and accept the warning. Now you'll be able to see what's causing the tests to hang (e.g. a prompt or alert that hasn't been dismissed).

### Adding a new e2e test

Open `web/src/e2e/index.e2e.test.tsx` and add a new `test`:

```TypeScript
        test.only('widgetizes quuxinators', async () => {
            await page.goto(baseURL + '/quuxinator/widgetize')
            await page.waitForSelector('.widgetize', { visible: true })
            // ...
        })
```

The full [Puppeteer API](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md) is quite large, but most tests only use a few common commands:

- `await page.goto(baseURL + '/some/route')` navigate to a URL
- `await page.waitForSelector(selector, { visible: true })` wait for an element to appear
- `await page.click(selector)` click on an element (must be visible, but not necessarily in the viewport)

> The easiest way to write CSS selectors is to inspect the element in your browser and look at the CSS classes. From there, you can write a selector and get immediate feedback:

Reviewer: I would like to document here that all selectors used in e2e tests should always be `e2e-` class names. I.e. if an element doesn't have such a class, one should be added (instead of using existing classes on the element). The e2e classes do not follow BEM.

Proposal:
```suggestion
CSS selectors in e2e tests should always refer to CSS classes prefixed with `e2e-`. This makes them easy to spot in the implementation and therefor less likely to accidentally break. `e2e-` classes are never referenced in stylesheets, they are added _in addition_ to styling classes. If an element you are trying to select does not have an `e2e-` class yet, modify the implementation to add it. You can see the classes of an element in the inspector:
```

![image](https://user-images.githubusercontent.com/1387653/54873834-6c605400-4d9c-11e9-823b-faa8871df395.png)

> Sometimes it's easier to add `data-e2e-*` attributes to the web app code than to construct a monstrous CSS or XPath selector to match an element. Clicking on a button corresponding to a particular element in a list is especially difficult, but easy with a `data-e2e-*` attribute:

Reviewer: ```suggestion
If the element you are trying to select appears multiple times on the page (e.g. a button in a list) and you need to select a specific instance, you can use `data-e2e-*` attributes in the implementation:
```

```HTML
<div data-e2e-item-name={this.props.name}>
  <span>{this.props.name}</span>

>   <button>Delete</span>

Reviewer: ```suggestion
  <button className="e2e-item-delete-button">Delete</button>
```

</div>
```

> Then you can select the button with `div[data-e2e-item-name="foo"] .button`.

Reviewer: ```suggestion
Then you can select the button with `[data-e2e-item-name="foo"] .e2e-item-delete-button`.
```

### E2e tests vs unit tests

In comparison to unit tests, e2e tests are slower and flakier but often more convenient.

E2e tests are typically beneficial for testing a happy-path through a user flow, such as adding a repository then running a search.

> E2e tests are probably not worth the slowness/flakiness for testing a matrix of inputs, benchmarking, checking DOM structure, or verifying the correctness of logic.

Reviewer: I think this suggestion should be included here:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/doc/dev/web_app.md#L101

## Visual snapshot tests

### Reviewing visual changes in a PR

When you submit a PR, a check from https://percy.io/Sourcegraph/Sourcegraph will appear:

![image](https://user-images.githubusercontent.com/1387653/54873096-ac1f3f80-4d8c-11e9-93ff-377b28121df4.png)

If Percy failed CI ❌ then click on the **Details** link to review the visual changes:

![image](https://user-images.githubusercontent.com/1387653/54873144-c0177100-4d8d-11e9-839c-3e344a6d872b.jpg)

Click the image on the right to toggle between diff and full image mode to review the change. Diff mode shows the changes in red.

If the changes are intended, click **Approve** 👍

Once you approve all of the changes, the Percy check will turn green ✅

### Adding a new visual snapshot test

Open `web/src/e2e/index.e2e.test.tsx` and add a new e2e test:

```TypeScript
        test('Repositories list', async () => {
            await page.goto(baseURL + '/site-admin/repositories?query=gorilla%2Fmux')

>             await page.waitForSelector('a[href="/github.com/gorilla/mux"]', { visible: true })

Reviewer: ```suggestion
            await page.waitForSelector('[e2e-repository-name="/github.com/gorilla/mux"]', { visible: true })
```
